### ISSUE TITLE
importKerasSequentialModelAndWeights seems stuck

### ISSUE THREAD

**Issue by himanshusinha1** [deeplearning4j/deeplearning4j#9180]
#### importKerasSequentialModelAndWeights load model stuck/hang

Hi, 

we noticed an issue in loading the h5 model using importKerasSequentialModelAndWeights and get stuck, It occurring in some cases mostly the model is importing fine but when it's stuck, it just stuck and take forever!

I tried sample code to reproduce the issue when we faced this issue in our Prod environment and the sample code is as below:

Note, below the loop, I just used to reproduce this issue. In Real env, we just used to import h5 model object once and get the prediction.

import org.deeplearning4j.nn.modelimport.keras.KerasModelImport;
import org.deeplearning4j.nn.modelimport.keras.exceptions.InvalidKerasConfigurationException;
import org.deeplearning4j.nn.modelimport.keras.exceptions.UnsupportedKerasConfigurationException;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;

public class Poc {

	public static void main(String[] args) throws Exception {
		
		final Logger LOGGER = LoggerFactory.getLogger(Poc.class);
		MultiLayerNetwork model = null;
		KerasModelImport OBJ = null;

		for(int i=1;i<=200;i++){

			try {

				String modelpath = "test.model.h5";

				LOGGER.debug("Calling importKerasSequentialModelAndWeights() for " + modelpath);

				model = OBJ.importKerasSequentialModelAndWeights(modelpath, false);
                                 // call predict to get the prediction
                                 //model.predict()

				LOGGER.debug("Imported importKerasSequentialModelAndWeights() for " + modelpath);

		     } catch (IOException e) {

				LOGGER.error(e.getMessage(), e);

			} catch (InvalidKerasConfigurationException e) {

				LOGGER.error(e.getMessage(), e);

			} catch (UnsupportedKerasConfigurationException e) {
				LOGGER.error(e.getMessage(), e);
		}
		}
}
}

* Deeplearning4j 1.0.0-beta6

Let me know if anyone encounters the same issue/ or if anyone gets a feasible solution to fix this.

**Comment by agibsonccc** [comment]
@himanshusinha1 I would consider not running the keras import inline. Consider running import and then directly using the ModelSerialize.restore(..) methods. If you can give me a network, I'm happy to profile it.

**Comment by himanshusinha1** [comment]
Hi @agibsonccc , Thanks, can you please describe how u mean by "import and then directly using the ModelSerialize.restore(..) methods"?

**Comment by agibsonccc** [comment]
@himanshusinha1  I met just convert the model via import and save the result, then load it in production.  The "normal" saving of models can be found here: https://deeplearning4j.konduit.ai/models/model-persistence

A dl4j computation graph maps to a functional model and a sequential is a multi layer network.

Beyond that, going forward I will encourage that for all model conversions. We're releasing a rewrite of the model import for tensorflow and new support for onnx that keras will be converted to later.

**Comment by himanshusinha1** [comment]
Hi @agibsonccc , Thanks, I am using the https://deeplearning4j.org/api/latest/org/deeplearning4j/nn/modelimport/keras/KerasModelImport.html as the model we developed is Keras based in python and saving as h5 than using in Java with dl4j as an example posted in the description section.
what will be the advantage to convert/save the model result? as KerasModelImport.importKerasSequentialModelAndWeights() providing the same feature correct? for us, it is working perfectly fine the way we use but we noticed sometimes while importing the model  h5 file using importKerasSequentialModelAndWeights gets in hung state and no response/return from importKerasSequentialModelAndWeights function. so seems like some issue with importKerasSequentialModelAndWeights function when loading the h5 file.

Let me know what could be issue thanks for your support in advance.

**Comment by agibsonccc** [comment]
@himanshusinha1  conversion from one framework to another will always have some overhead. I'm aware it's convenient to do it the way you are. We'll discourage that going forward for our other frameworks due to edge cases like what you see with keras models on mobile. 
Hdf5 doesn't produce bindings for ARM so our keras users have to run the conversion on desktop.

Often the formats from python involve c code which may not be supported on various platforms. A tensor from tensorflow or a keras model will have a different format than the ones that are actually used in memory.

While we'll do our best to minimize that overhead (and I certainly will try to work on your problem if you give me the model like I asked) it's still never going to be fast as something we can completely control. Conversion of keras models sometimes involves 1 off code. A good example is lambdas. You may need to change your lambda function if you want it to work properly.  Maintaining that kind of code in production can cause a bit of overhead depending on what you're doing.

When you decouple your conversion workflow from your deployment one it allows you to just produce a model that will work more reliably.

**Comment by agibsonccc** [comment]
Closing due to lack of activity.

**Comment by himanshusinha1** [comment]
Thanks for your support @agibsonccc, we tried the soln you proposed, and it's working really well!
